packages/edit-site/src/components/add-new-template/new-template.js

```diff
@@ -128,64 +135,135 @@ export default function NewTemplate( { postType } ) {
 			} );
 		}
 	}
-
-	const existingTemplateSlugs = map( templates, 'slug' );
-
+	const existingTemplateSlugs = ( existingTemplates || [] ).map(
+		( { slug } ) => slug
+	);
 	const missingTemplates = filter(
 		defaultTemplateTypes,
 		( template ) =>
 			includes( DEFAULT_TEMPLATE_SLUGS, template.slug ) &&
 			! includes( existingTemplateSlugs, template.slug )
 	);
 
-	if ( ! missingTemplates.length ) {
+	const extraTemplates = ( postTypes || [] ).reduce(
+		( accumulator, _postType ) => {
+			const { slug, labels, icon } = _postType;
+			const hasGeneralTemplate = existingTemplateSlugs?.includes(
+				`single-${ slug }`
+			);
+			const hasEntities = postTypesEntitiesInfo?.[ slug ]?.hasEntities;
+			const menuItem = {
+				slug: `single-${ slug }`,
+				title: sprintf(
+					// translators: %s: Name of the post type e.g: "Post".
+					__( 'Single item: %s' ),
+					labels.singular_name
+				),
+				description: sprintf(
+					// translators: %s: Name of the post type e.g: "Post".
+					__( 'Displays a single item: %s.' ),
+					labels.singular_name
+				),
+				// `icon` is the `menu_icon` property of a post type. We
+				// only handle `dashicons` for now, even if the `menu_icon`
+				// also supports urls and svg as values.
+				icon: icon?.startsWith( 'dashicons-' )
+					? icon.slice( 10 )
+					: null,
+			};
+			// We have a different template creation flow only if they have entities.
+			if ( hasEntities ) {
+				menuItem.onClick = ( template ) => {
+					setShowCustomTemplateModal( true );
+					setEntityForSuggestions( {
+						type: 'postType',
+						slug,
+						labels,
+						hasGeneralTemplate,
+						template,
+						postsToExclude:
+							postTypesEntitiesInfo[ slug ].existingPosts,
+					} );
+				};
+			}
+			// We don't need to add the menu item if there are no
+			// entities and the general template exists.
+			if ( ! hasGeneralTemplate || hasEntities ) {
+				accumulator.push( menuItem );
+			}
+			return accumulator;
+		},
+		[]
+	);
+	if ( ! missingTemplates.length && ! extraTemplates.length ) {
 		return null;
 	}
-
 	// Update the sort order to match the DEFAULT_TEMPLATE_SLUGS order.
-	missingTemplates.sort( ( template1, template2 ) => {
+	// TODO: check sorting with new items.
```

Comment:
What to do here?

Reply:
I'll handle this in the next iterations. Currently we append new items in the end, but we will revisit when more items are added.